Draft: Add PythonSoftIOC record creation directly from PandA

AlexanderWells-diamond opened the pull request:
Note this is very much a work in progress, and it currently isn't functional.

AlexanderWells-diamond commented on pandablocks/commands.py:
```diff
@@ -298,54 +334,301 @@ class GetFieldInfo(Command[Dict[str, FieldInfo]]):
+                    [str, str, Optional[str]],
+                    Tuple[FieldInfo, List[_FieldCommandMapping]],
+                ],
+            ] = {
```
mypy doesn't seem to be that smart - you end up with this lovely inferred type:
```
(variable) _commands_map: dict[tuple[Literal['time'], None] | tuple[Literal['bit_out'], None] | tuple[Literal['pos_out'], None] | tuple[Literal['ext_out'], Literal['timestamp']] | tuple[Literal['ext_out'], Literal['samples']] | tuple[Literal['ext_out'], Literal['bits']] | tuple[Literal['bit_mux'], None] | tuple[Literal['pos_mux'], None] | tuple[Literal['param'], Literal['uint']] | tuple[Literal['read'], Literal['uint']] | tuple[Literal['write'], Literal['uint']] | tuple[Literal['param'], Literal['scalar']] | tuple[Literal['read'], Literal['scalar']] | tuple[Literal['write'], Literal['scalar']] | tuple[Literal['param'], Literal['time']] | tuple[Literal['read'], Literal['time']] | tuple[Literal['write'], Literal['time']] | tuple[Literal['param'], Literal['enum']] | tuple[Literal['read'], Literal['enum']] | tuple[Literal['write'], Literal['enum']], Any]
```

So I'll leave the type specification in.

AlexanderWells-diamond commented on pandablocks/commands.py:
```diff
@@ -298,54 +334,301 @@ class GetFieldInfo(Command[Dict[str, FieldInfo]]):
+                ("write", "enum"): self._commands_enum,
+            }
+
+            if (field_type, subtype) in _commands_map:
+                # Create type-specific commands
+                field_info, command_mapping = _commands_map[(field_type, subtype)](
+                    name, field_type, subtype
+                )
```
Tweaked and applied.

AlexanderWells-diamond commented on pandablocks/commands.py:
```diff
@@ -298,54 +334,301 @@ class GetFieldInfo(Command[Dict[str, FieldInfo]]):
+            # hence why it is handed here.
+            # Note that we don't get the description for any attributes - these are
+            # fixed strings and so not worth retrieving dynamically.
+            if not self.skip_description:
```
When this flag is used to disable retrieving attributes, do you have any preference as to whether we return a base `FieldInfo` class, or a more specific subclass (e.g. `BitMuxFieldInfo`) but with only the `type` and `subtype` attributes set?

thomascobb replied:
I'd say we only return the basic `FieldInfo`

AlexanderWells-diamond replied:
Done in #5ad6cb3.

AlexanderWells-diamond commented on pandablocks/commands.py:
```diff
@@ -298,54 +334,301 @@ class GetFieldInfo(Command[Dict[str, FieldInfo]]):
+        return (
+            field_info,
+            [
+                _FieldCommandMapping(
```
Implemented a much-adapted version of this suggestion. Couldn't directly call `_execute_commands` as that causes the function to yield all the way back up to the connection, thus invoking a round trip. So I handled the generator usage myself with the `next()` and `send()` methods.

This change could probably do with a review all of its own.

AlexanderWells-diamond commented on examples/_ioc.py:
```diff
@@ -0,0 +1,1017 @@
+    changes = returned_infos[-1]
+    if changes.in_error:
+        raise Exception("TODO: Some better error handling")
+    # TODO: Do something with changes.no_value ?
```
Done in https://github.com/PandABlocks/PandABlocks-client/pull/28/commits/017415ec72777a37913386e7acdfb8a1e598a2cf

AlexanderWells-diamond commented:
> As it is a single command you should be able to write:
> 
> ```python
> max = yield from Get(f"{self.block}1.{field_name}.MAX").execute()
> ```
Fixed in https://github.com/PandABlocks/PandABlocks-client/pull/28/commits/f79514e395621133a705de9178fac282d0d97734

AlexanderWells-diamond commented:
@thomascobb I think this is in a good state to review again. Let me know if you'd prefer me to open a new PR or anything to make the history cleaner.

There's still a few things to do, which I think most need some amount of discussion with you:

- Warning/Error/Exception reporting mechanism - there's currently a combination of `print()`, `assert` and `TODO` scattered around
- The scalar record equivalents for some TABLE records - especially the enum ones as the labels are currently inaccessible
- Testing of the `ioc.py` file (the `commands.py` changes I think are well tested)
- Many TODOs, some of which require discussion

thomascobb commented on pandablocks/commands.py:
```diff
@@ -406,30 +773,77 @@ class GetChanges(Command[Changes]):
+                    field = line[0:-1]
+
+                    # The METADATA tables are strings, and cannot be requested in Base64
+                    if self.multiline_base64 and "*METADATA" not in field:
```
Although you don't seem to use it anywhere, so maybe we should ditch the base64 argument here?

AlexanderWells-diamond replied:
This was written when we were considering using Base64 encoding for all table values. We don't use it anymore. I thought I'd leave it in as I'd written it, but if its preferred to delete it I can do that.

thomascobb replied:
Lets delete it

AlexanderWells-diamond replied:
Deleted in https://github.com/PandABlocks/PandABlocks-client/pull/28/commits/fdcd060ecd55d5fbaca4fb1be2b87aaa85242ef2

AlexanderWells-diamond commented on pandablocks/commands.py:
```diff
@@ -465,11 +879,11 @@ def execute(self) -> ExchangeGenerator[List[str]]:
         for field in table.no_value:
```
That'll cause problems - the changes will be reset client-wide, so every use of `GetChanges` will be affected. For example in our usage of `GetChanges` we sleep for 1 second between each call. If a PandA change and a `*CHANGES=S` happens in the same sleep second, wouldn't the polling `GetChanges` then potentially lose the PandA change that occurred?

thomascobb replied:
Good point, we should just document that `GetState` and `GetChanges` interact with each other and should be used with care together

AlexanderWells-diamond replied:
I can do that - it'll also be easy to work around by just using two different clients, if needed. I avoided using multiple clients in my code.

Could I confirm that the `*CHANGES=S` request will need to be in its own `yield`, before the existing list in `_execute_commands`, as the requests in `_execute_commands` may not necessarily be processed in order? Or is it actually safe to just add the new command to the top of that list? Thanks.

thomascobb replied:
I should clarify, we shouldn't do the `*CHANGES=S` command here, just document that `GetChanges` and `GetState` interfere with each other and should be run in separate clients. (It's academic, but it should be safe to do them all in the same `_execute_commands` as the command side of the server is single threaded.

AlexanderWells-diamond replied:
Documentation note added in https://github.com/PandABlocks/PandABlocks-client/pull/28/commits/e029b5f93c5853598cb136761c033a2cc744e9af

AlexanderWells-diamond commented on pandablocks/ioc.py:
```diff
@@ -0,0 +1,1720 @@
+    if not block_name_number[-1].isdigit():
+        return block_and_field_name.replace(
+            block_name_number, block_name_number + "1", 1
+        )
```
Agreed. Your comment made me realise that the docs are a touch unclear - it expects ":"-delimited fields (not that it made a difference before your change)
Change made in https://github.com/PandABlocks/PandABlocks-client/pull/28/commits/7abc82fc095fa5fcc9583f2a97f3ac20679ab629

AlexanderWells-diamond commented on pandablocks/ioc.py:
```diff
@@ -0,0 +1,1720 @@
+        assert packed
+
+        # 2-D array -> 1-D array -> list[int] -> list[str]
+        return [str(x) for x in packed.flatten().tolist()]
```
I ended up doing everything without base64. I just had a talk to Michael and he thinks that in most cases base64 will be more efficient for data transfer, but in some cases it will be less efficient, but probably a benefit overall for all but empty tables. I could add in the base64 requests again if wanted?

thomascobb replied:
No, lets leave it without base64, the sequencer tables are half empty at the moment so that should be more efficient

AlexanderWells-diamond commented on pandablocks/ioc.py:
```diff
@@ -0,0 +1,1720 @@
+            changes = await client.send(GetChanges(ChangeGroup.ALL, True))
+            if changes.in_error:
+                pass
+                # TODO: Warning/error here
```
Thanks, I'll look at adding `logging.warning` or other appropriate calls everywhere I've said I need a warning/error message.

AlexanderWells-diamond commented:
This PR has now been updated for PythonSoftIOC 4.0.1, which fixed all of the blocking issues outlined in my previous comment. I also took the opportunity to tidy things up a bit.

AlexanderWells-diamond commented:
Updated for `softioc 4.1.0`. This adds Blocking to the records, so that a `caput` will only return once any `on_update` processing has completed. Tests modified to accommodate this, which actually makes them faster as I could remove all of the `sleep` calls I had to include previously.

coretl commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+            return {}
+
+    # Map a field's (type, subtype) to a function that creates and returns record(s)
+    _field_record_mapping: Dict[
```
Could you avoid a layer of `_make` functions if you put a few more things in this table? Like record factory, or number of values?

AlexanderWells-diamond replied:
Probably possible, I'll investigate...

coretl reviewed (commented):
I've only looked at ioc.py so far, I'll have a go at doing the PVA exposing of bit_muxes and then do a more thorough review

AlexanderWells-diamond commented on pandablocks/ioc/_tables.py:
```diff
@@ -0,0 +1,564 @@
+            # Mask to remove every bit that isn't in the range we want
+            mask = (1 << bit_len) - 1
+
+            # Can't use proper numpy types, that's only available in 1.21+
```
I don't quite understand what I did here. We're already on `1.21.6`, which is the last `numpy` version that supports python3.7. 

I'm reworking this section to be a bit better with types.

AlexanderWells-diamond replied:
Refactored in https://github.com/PandABlocks/PandABlocks-client/pull/28/commits/db34404b2262482dd92c8e47a2c83fc70d7f0914

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+        record_dict[units_record_name] = self._create_record_info(
+            units_record_name,
+            "Units of time setting",
+            builder.mbbIn,
```
Easy to change to mbbOut, and the framework should already be changing `<FIELD>.UNITS` in the PandA whenever the value itself is changed?

The more difficult one is changing the `EGU` value. After the IOC is initialised, the `<record>.EGU` field isn't accessible. Does this mean that I'll have to do a `caput` to it, despite the record being within the same process?

coretl replied:
You should be able to do `dbpf` to it

AlexanderWells-diamond replied:
Done in 35919d5bac0c99259411b586f17611234a325368 and 15ac1839dde4d3b4a38d8745cba2a5bbe3695118

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+        )
+
+        min_record_name = EpicsName(record_name + ":MIN")
+        record_dict[min_record_name] = self._create_record_info(
```
Not sure I follow - are you saying I just need to do `<record>.DRVL = field_info.min`? That'll be fine for setting the initial value, but will cause problems pushing/pulling values to/from PandA, as the architecture is not designed to be able to push/pull from record attributes. 

This is probably related to the questions about `.EGU` on the above comment thread.

AlexanderWells-diamond replied:
From what I can tell the MIN field is in the same units as the parent field:

```
< PULSE1.DELAY.UNITS=s
OK
< PULSE1.DELAY.MIN?
OK =8e-09
< PULSE1.DELAY.UNITS=ms
OK
< PULSE1.DELAY.MIN?
OK =8e-06
< PULSE1.DELAY.UNITS=us
OK
< PULSE1.DELAY.MIN?
OK =0.008
```
As you can see, the value automatically changes when the UNITS are specified. What I think this means is that I'll have to query the PandA for the new value of MIN, whenever the UNITS are changed, and then set the `DRVL` field of the parent record. 

I'm not terribly happy with having to do a network call in an `on_update` function, but at least now we have Blocking records.

AlexanderWells-diamond replied:
Done in 15ac1839dde4d3b4a38d8745cba2a5bbe3695118

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+        if record_creation_func in OUT_RECORD_FUNCTIONS:
+            # Ensure VAL is clamped to valid range of values.
+            # The DRVH field is a signed LONG value, but PandA uses unsigned 32-bit
+            # which can overflow it.
```
At what point do we do the conversion to int? And if the record itself is now an analog, rather than long, doesn't that not really reflect what the underlying data in the PandA actually is?

coretl replied:
If we use an `ao` with `PREC=0` then the value of the record will be a float64, but displayed as rounded to the nearest integer. This means that on the GUI it will look like the value came from a longin. A float64 can display up to 56-bit integers without rounding error, so we can convert to int in `on_update` before sending to PandA. This means we won't overflow and can ditch the clamping of max values code here

AlexanderWells-diamond replied:
Implemented in 3e883740c4e375a167da8ea0b27ea45375c3df9c. I left the bounds checking in place, although if I understand correctly its now sort of unnecessary as the float64 is large enough to hold any value the PandA might send.

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+            record_dict[record_name].record.DRVL = 0
+            record_dict[record_name].record.DRVH = min(field_info.max, 2147483647)
+
+        max_record_name = EpicsName(record_name + ":MAX")
```
This is another case where I think there are significant architectural changes required to do this change, assuming MAX is a value that could be changed?

AlexanderWells-diamond replied:
Fixed in 0f917757ca435313cc3c86990fd5c05341519d91

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+            # which can overflow it.
+            assert field_info.max
+            record_dict[record_name].record.DRVL = 0
+            record_dict[record_name].record.DRVH = min(field_info.max, 2147483647)
```
That value is twice what I had written:
```
>>> np.iinfo(np.uint32).max
4294967295
```
So I guess I made a mistake and used an `int32` when I originally wrote this.

coretl replied:
I think the `longin` is an `int32`, my comment was about using an `ai` which is `float64` which gives 56 bits of integer position, but then we need to use `DRVL` and `DRVH` to set bounds to it and `PREC=0` to indicate to the GUI to disregard fractional values, probably with a `validate` function to disregard on put

AlexanderWells-diamond replied:
Done in 02f5c11d7fac31814710dffe5a6818b5a68cf061

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+            initial_value=field_info.scale,
+        )
+
+        units_record_name = EpicsName(record_name + ":UNITS")
```
Another architectural problem point.

AlexanderWells-diamond replied:
I can do this fairly easily, in a similar way to the EGU and DRVL fields for field type TIME, except for one problem: The `pos_out` record type is  `builder.longIn`, which means `DISP=1`. This blocks all put'ting, even from `dbPutField`. So I cannot easily update `EGU` when the `UNITS` change.

The only way around this I can see is to unset `DISP`, do the put to `EGU`, and then re-set `DISP`. This seems far from ideal.

On a related note, it seems odd that the `UNITS` are settable but the value itself is not. Looks like that is how PandA works though, so we'll just reflect that here.

coretl replied:
On reflection this sounds fine. Anything with user-settable units (like `time` or `param time`) already has those units on the GUI, so there's not point putting them in the EGU field as well

AlexanderWells-diamond commented:
While working on the `DRVL` value of `time` fields, I discovered that the `MIN` value is not reported via `*CHANGES?`. I've raised that as an issue here: https://github.com/PandABlocks/PandABlocks-server/issues/30

However, as we won't deploy any potential fixes to all PandAs, I'll have to work with the code as-is. This'll mean our polling code will have to perform an explicit query for the new value of `MIN`, when it sees the `UNITS` field has updated.

coretl commented:
> However, as we won't deploy any potential fixes to all PandAs, I'll have to work with the code as-is. This'll mean our polling code will have to perform an explicit query for the new value of `MIN`, when it sees the `UNITS` field has updated.

Is the `MIN` value is scaled with the units? If so, we could convert from its value in initial units to the requested units, then we wouldn't need any changes to the server.

AlexanderWells-diamond commented:
Yes, `MIN` is scaled with the units. I did pose the idea of calculating it client-side to Michael, but it does mean that we'd be duplicating code from the server. I'm also not sure exactly how the units are defined - if it can only ever be min, s, ms, and us, we _could_ calculate it ourselves. But as soon as that list changes it becomes significantly harder to do the conversion ourselves.

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1756 @@
+        client, dispatcher, record_prefix
+    )
+
+    asyncio.create_task(update(client, all_records, 1, all_values_dict))
```
Done in 8d7ecd4436b1b76c9a8e0958a1443cae082086f0

AlexanderWells-diamond commented:
> > ```
> > * Again on the `FieldInfo` class: The fields of the child classes are all optional. This is due to the `dataclass` restriction that once there is an optional field (the description in the base `FieldInfo` may not exist) all subsequently defined fields must also be optional. However we know that these fields must exist, and this leads to a large amount of `assert` statements throughout the code to make `mypy` happy. The solution I'm aware of is to split the classes into smaller dataclasses, grouping together the required and the optional parameters, and then stitch them together on a class-by-class basis to create the right order of parameters. This would mean the right fields would be required, but it would also make it a lot harder to inspect a class and know what fields were in it as they'd be inherited from two or three different classes.
> > ```
> 
> Can this be solved by making all the base class fields required, then passing them a junk value? Otherwise we end up with something like https://stackoverflow.com/questions/51575931/class-inheritance-in-python-3-7-dataclasses/73095470#73095470 which doesn't actually help with the mypy issue

I must have misunderstood what was causing me the problem originally: It wasn't the `Optional`, it was actually the default value. Removing them from the base FieldInfo allowed me to remove all the incorrect `Optional` markers in all the subclasses. Done in 9609161e3873ee80f9732bbc4324ffe6fa3ffaf7 and df7bdd8ed943c5eb7b1c39d340771bffc4115800.

> 
> > ```
> > * Due to PandA reporting any changes we do through `Put` commands back to us through `GetChanges`, we end up setting record values twice. I'm unsure of the full implications of this - I disable processing for Out records, and In records don't have `validate` or `on_update` methods anyway, but tools like `camonitor` may report changes, especially if any records have `always_update` set. This may also pose issues for fast-changing records, as the polling in `GetChanges` will only happen periodically and so could easily miss updates.
> > ```
> 
> I do this at the moment by setting the a flag that ignores the next update if the new value is equal to the value we set: https://github.com/dls-controls/pymalcolm/blob/a9caeccf5b75c444f653f6785edef27599e4b60b/malcolm/modules/pandablocks/parts/pandafieldpart.py#L60-L75

Implemented in adf271313e58c4a9f9d192c8e52f5dc4abad1bc3. This required a fair chunk of re-architecting, to allow access to the `RecordInfo` structure in both the EPICS on-update callback and the processing of `*CHANGES?` requests. It's pretty ugly, but is the simplest way to achieve the required result.

> 
> > ```
> > * Lastly on logging, on my test PandA we see a lot of warnings due to label lengths exceeding what can be stored in an `mbbi/mbbo` record. These are all `warning` level log messages. We did discuss a similar thing for the `description` truncation messages, which have been bumped down to the `info` log level. A possible thing to do here is to use `string` records, rather than `mbbi/mbbo` records, and apply the [validator class](https://github.com/PandABlocks/PandABlocks-client/blob/pythonsoftioc/pandablocks/ioc/ioc.py#L359) I wrote for them.
> > ```
> 
> I think we should change all the read-only enums to `string` records regardless off their length, which we can wrap in as a PVA enum later. That will catch most of them.

I changed the read-only enum (for param-read ) with a string record. All other record types are `mbbOut`, and so cannot so easily be changed. See 193dbd89faeaa5f45b23b35f6689f7511c1a6af9.

> > ```
> > * The [_create_record_info](https://github.com/PandABlocks/PandABlocks-client/blob/pythonsoftioc/pandablocks/ioc/ioc.py#L448) method might be doing far too much at this point - it tries to cover all cases for all possible `builder` record creation functions. Any thoughts on whether it needs refactoring?
> > ```
> 
> I mentioned in the review comments, might be good to expand the table to have a few more arguments which might help here.

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1882 @@
+
+        updater: _WriteRecordUpdater
+
+        record_info = self._create_record_info(
+            record_name,
+            field_info.description,
+            builder.Action,
+            int,  # not bool, as that'll treat string "0" as true
+            ZNAM=ZNAM_STR,
+            ONAM=ONAM_STR,
+            on_update=lambda v: updater.update(v),
+        )
```
To confirm, do you expect both of those values to generate the same PandA command, e.g. `FOO.BAR=` ?

coretl replied:
yes

AlexanderWells-diamond replied:
Implemented in e9d0140290d208ed50b7b781ce9872cea4013a1e

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1882 @@
+            record_name,
+            field_info.description,
+            builder.boolIn,
+            int,
```
Implemented in b8b5770f43b6dafe0d46496a4fa969ae4f5c8874

AlexanderWells-diamond commented on pandablocks/ioc/ioc.py:
```diff
@@ -0,0 +1,1882 @@
+        raise RuntimeError("Unexpected state - softioc task already exists")
+
+    create_softioc_task = asyncio.create_task(
+        update(client, all_records, 1, all_values_dict)
```
Implemented in 203f1c903d13061ea0e536759a13d154b9904552

AlexanderWells-diamond commented:
@coretl Think I'm done with review comments again.
